pyccl/halos/profiles_cib.py

```diff
@@ -189,21 +189,19 @@ def _Lumcen(self, M, a):
         return Lumcen
 
     def _Lumsat(self, M, a):
-        Lumsat = np.zeros_like(M)
-        # Loop over Mparent
-        # TODO: if this is too slow we could move it to C
-        # and parallelize
-        for iM, Mparent in enumerate(M):
-            if Mparent > self.Mmin:
-                # Array of Msubs (log-spaced with 10 samples per dex)
-                nm = max(2, int(np.log10(Mparent/1E10)*10))
-                msub = np.geomspace(1E10, Mparent, nm+1)
-                # Sample integrand
-                dnsubdlnm = self.dNsub_dlnM_TinkerWetzel10(msub, Mparent)
-                Lum = self._Lum(np.log10(msub), a)
-                integ = dnsubdlnm*Lum
-                Lumsat[iM] = simps(integ, x=np.log(msub))
-        return Lumsat
+        res = np.zeros_like(M)
+        M_use = M[M >= self.Mmin, None]
+        logM = np.log10(M_use)
+        LOGM_MIN = 10
+        nm = max(2, 10*int(np.max(logM) - LOGM_MIN))
+        msub = np.linspace(LOGM_MIN, np.max(logM), nm+1)[None, :]
+
+        Lum = self._Lum(msub, a)
+        dnsubdlnm = self.dNsub_dlnM_TinkerWetzel10(10**msub, M_use)
+        integ = dnsubdlnm * Lum
+        Lumsat = simps(integ, x=np.log(10)*msub)
```

Comment:
I am not 100% sure this is doing the same calculation. Note that the upper limit of the integral being calculated in the old loop changes with each element of the loop (Mparent). Here, the upper limit is always the same.

Reply:
Yes, it's not the same, but we can exploit the fact that the subhalo mass function drops extremely rapidly after `Msub > Mparent`, so the high-mass contributions are next to zero.
To be precise, `Lum` changes by some orders of magnitude as mass increases, but `dnsubdlnm` instantly drops 8 orders of magnitude for a mass change of 0.1 dex, which is the sampling rate either way.
So unless I didn't read it right, the high-mass contributions are negligible.